### PR TITLE
(opt): cache trivia green ids in the parser

### PR DESCRIPTION
**Pull request by orizi**
Most terminals carry one of a handful of leading/trailing trivia lists
(whitespace, newline+indentation), yet each terminal re-interned its
trivia into the global green table - hash, shard lock, and structural
equality against a huge map, which under the parallel warmup became a
contention point (sched_yield inside interning was the top parsing cost).
A parser-local cache serves repeats without touching the global table;
sound since green ids are immortal.

Improves staking diagnostics by ~9%; corelib is unaffected (its trivia
diversity is higher relative to file count).

Co-Authored-By: Claude Fable 5 <noreply@anthropic.com>

### CODE REVIEW

**Comment by orizi** [comment]
* **#10343** <a href="https://app.graphite.com/github/pr/starkware-libs/cairo/10343?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/starkware-libs/cairo/10343?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by orizi — COMMENTED** [review]
> @orizi made 2 comments.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/10343)* status: 0 of 5 files reviewed, 2 unresolved discussions.

___
*[`crates/cairo-lang-parser/src/parser.rs` line 3737 at r2](https://reviewable.io/reviews/starkware-libs/cairo/10343#-Ozq8_Ye466hLDdpud6w:-Ozq8_Ye466hLDdpud6x:b-7ce9eu) ([raw file](https://github.com/starkware-libs/cairo/blob/45b6a730abff7678c18acc21df82d16907cbeac3/crates/cairo-lang-parser/src/parser.rs#L3737)):*
> ```XML
>         self.consume_pending_skipped_diagnostics();
> 
>         let leading_width = trivia_total_width(self.db, &leading_trivia);
> ```



the usage `trivia_total_width` sounds like it should be a part of `.trivia_green`.
___
*[`crates/cairo-lang-parser/src/parser.rs` line 3744 at r2](https://reviewable.io/reviews/starkware-libs/cairo/10343#-Ozq80_h6RXq0NcW3X0x:-Ozq80_h6RXq0NcW3X0y:b-d4hmn6) ([raw file](https://github.com/starkware-libs/cairo/blob/45b6a730abff7678c18acc21df82d16907cbeac3/crates/cairo-lang-parser/src/parser.rs#L3744)):*
> ```XML
>             new_leading_trivia.extend(leading_trivia);
>             Trivia::new_green(self.db, &new_leading_trivia)
>         };
> ```

shouldn't pending trivia be part of the basic grouping?
and be cached as well?



_Code quote:_
```XML
        let leading_width = trivia_total_width(self.db, &leading_trivia);
        let leading_end = terminal_start.add_width(leading_width);
        let leading = if new_leading_trivia.is_empty() {
            self.trivia_green(&leading_trivia, TextSpan::new(terminal_start, leading_end))
        } else {
            new_leading_trivia.extend(leading_trivia);
            Trivia::new_green(self.db, &new_leading_trivia)
        };
```

**Review by orizi — COMMENTED** [review]
> @orizi+AGNT made 2 comments and resolved 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/10343)* status: 0 of 5 files reviewed, all discussions resolved.

___
*[`crates/cairo-lang-parser/src/parser.rs` line 3737 at r2](https://reviewable.io/reviews/starkware-libs/cairo/10343#-Ozq8_Ye466hLDdpud6w:-OzqFyz_9PCF2f7kOukV:bvcl1o9) ([raw file](https://github.com/starkware-libs/cairo/blob/45b6a730abff7678c18acc21df82d16907cbeac3/crates/cairo-lang-parser/src/parser.rs#L3737)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

the usage `trivia_total_width` sounds like it should be a part of `.trivia_green`.
</blockquote></details>

Done - `trivia_green` now takes just the start offset and computes the width itself; the caller derives the trailing start from the returned green's width.
___
*[`crates/cairo-lang-parser/src/parser.rs` line 3744 at r2](https://reviewable.io/reviews/starkware-libs/cairo/10343#-Ozq80_h6RXq0NcW3X0x:-OzqFy_j2m1kFXmYFtWp:b-5z0e2w) ([raw file](https://github.com/starkware-libs/cairo/blob/45b6a730abff7678c18acc21df82d16907cbeac3/crates/cairo-lang-parser/src/parser.rs#L3744)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

shouldn't pending trivia be part of the basic grouping?
and be cached as well?
</blockquote></details>

Done - pending trivia now flows through the same `trivia_green` grouping (it is source-contiguous ending at `terminal_start`, so its start is derived by subtracting its width - no new state), making it cache-eligible like any other list.

**Review by orizi — COMMENTED** [review]
> @orizi made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/10343)* status: 0 of 5 files reviewed, 1 unresolved discussion.

___
*[`crates/cairo-lang-parser/src/lexer.rs` line 67 at r3](https://reviewable.io/reviews/starkware-libs/cairo/10343#-OzqnPjd8Bq27q90dn4q:-OzqnPjd8Bq27q90dn4r:b-dlg0p0) ([raw file](https://github.com/starkware-libs/cairo/blob/100272dd662a324f142ca94c18654ccb8d25514e/crates/cairo-lang-parser/src/lexer.rs#L67)):*
> ```XML
>         &mut self,
>         db: &'db dyn Database,
>         leading: bool,
> ```

already enough - according to the usages.



_Suggestion:_
```XML
    fn match_trivia(
        &mut self,
        db: &'a dyn Database,
        leading: bool,
```

**Review by orizi — COMMENTED** [review]
> @orizi made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/10343)* status: 0 of 5 files reviewed, 2 unresolved discussions.

___
*[`crates/cairo-lang-parser/src/parser.rs` line 3759 at r3](https://reviewable.io/reviews/starkware-libs/cairo/10343#-Ozqnk5OC9EJhhn7HeVE:-Ozqnk5OC9EJhhn7HeVF:beu6pv6) ([raw file](https://github.com/starkware-libs/cairo/blob/100272dd662a324f142ca94c18654ccb8d25514e/crates/cairo-lang-parser/src/parser.rs#L3759)):*
> ```XML
>         let end = start.add_width(trivia_total_width(self.db, trivia));
>         let text = TextSpan::new(start, end).take(self.text);
>         if text.bytes().all(|b| matches!(b, b' ' | b'\r' | b'\t' | b'\n')) {
> ```

isn't there a "is whitespace" sort of fn?

**Review by orizi — COMMENTED** [review]
> @orizi+AGNT made 2 comments and resolved 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/10343)* status: 0 of 5 files reviewed, all discussions resolved.

___
*[`crates/cairo-lang-parser/src/lexer.rs` line 67 at r3](https://reviewable.io/reviews/starkware-libs/cairo/10343#-OzqnPjd8Bq27q90dn4q:-OzqoaWa-qNXTn4kRtxL:b-8lwl33) ([raw file](https://github.com/starkware-libs/cairo/blob/100272dd662a324f142ca94c18654ccb8d25514e/crates/cairo-lang-parser/src/lexer.rs#L67)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

already enough - according to the usages.
</blockquote></details>

Done - unified all the lexer methods (and `match_terminal`) on the struct lifetime; `tokenize_all` in the test now takes `&'db str` accordingly.
___
*[`crates/cairo-lang-parser/src/parser.rs` line 3759 at r3](https://reviewable.io/reviews/starkware-libs/cairo/10343#-Ozqnk5OC9EJhhn7HeVE:-Ozqob4lCFP2m_IwZqZv:b-o3q4he) ([raw file](https://github.com/starkware-libs/cairo/blob/100272dd662a324f142ca94c18654ccb8d25514e/crates/cairo-lang-parser/src/parser.rs#L3759)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

isn't there a "is whitespace" sort of fn?
</blockquote></details>

Done - `u8::is_ascii_whitespace`. (It additionally admits form feed, which the lexer never emits as trivia, so eligibility is unchanged.)

**Review by orizi — COMMENTED** [review]
> @orizi made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/10343)* status: 0 of 5 files reviewed, 1 unresolved discussion.

___
*[`crates/cairo-lang-parser/src/parser.rs` line 3751 at r4](https://reviewable.io/reviews/starkware-libs/cairo/10343#-OzqqKJxE7Rw3KlVHOM5:-OzqqKJxE7Rw3KlVHOM6:bqwwc0y) ([raw file](https://github.com/starkware-libs/cairo/blob/206cbd97430b31a7bdbcd7ea62e8a68efb8467cf/crates/cairo-lang-parser/src/parser.rs#L3751)):*
> ```XML
>         let trailing = self.trivia_green(&trailing_trivia, trailing_start);
>         Terminal::new_green(self.db, leading, token, trailing)
>     }
> ```

or something of that sort.



_Suggestion:_
```XML
        let LexerTerminal { text, kind: _, leading_trivia, trailing_trivia } = lexer_terminal;
        let token = Terminal::TokenType::new_green(self.db, text);
        let mut pending_trivia = mem::take(&mut self.pending_trivia);

        self.consume_pending_skipped_diagnostics();

        // Pending trivia is source-contiguous and ends exactly at `terminal_start`.
        let leading_start = terminal_start.sub_width(trivia_total_width(self.db, &pending_trivia));
        let trailing_start = terminal_start.add_width(trivia_total_width(self.db, &leading_trivia)).add_width(TextWidth::from_str(text.long(self.db)));
        let leading_trivia = if pending_trivia.is_empty() {
            leading_trivia
        } else {
            pending_trivia.extend(leading_trivia);
            pending_trivia
        };
        Terminal::new_green(self.db, self.trivia_green(&leading_trivia, leading_start), token, self.trivia_green(&trailing_trivia, trailing_start))
    }
```

**Review by orizi — COMMENTED** [review]
> @orizi+AGNT made 1 comment and resolved 1 discussion.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/10343)* status: 0 of 5 files reviewed, all discussions resolved.

___
*[`crates/cairo-lang-parser/src/parser.rs` line 3751 at r4](https://reviewable.io/reviews/starkware-libs/cairo/10343#-OzqqKJxE7Rw3KlVHOM5:-Ozqr9ql8d_212y7ojeY:b-pt8np2) ([raw file](https://github.com/starkware-libs/cairo/blob/206cbd97430b31a7bdbcd7ea62e8a68efb8467cf/crates/cairo-lang-parser/src/parser.rs#L3751)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

or something of that sort.
</blockquote></details>

Done - applied as suggested.

**Review by eytan-starkware — COMMENTED** [review]
> @eytan-starkware+AGNT made 6 comments.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/10343)* status: 0 of 5 files reviewed, 6 unresolved discussions (waiting on eytan-starkware, orizi, and TomerStarkware).

___
*[a discussion](https://reviewable.io/reviews/starkware-libs/cairo/10343#-P-97WWa3OGqm5qhL8ML:-P-97WWb3Z_DEDy2r6Ws:bnabdgl) (no related file):*
Note: the comments below are from an automatic orizi-review run (Claude agents reviewing in Ori's style, findings adversarially verified before posting). Treat with the usual bot skepticism.
___
*[`crates/cairo-lang-parser/src/parser.rs` line 76 at r5](https://reviewable.io/reviews/starkware-libs/cairo/10343#-P-941vm3VUGS6fZMCnb:-P-941vm3VUGS6fZMCnc:bnjzhxe) ([raw file](https://github.com/starkware-libs/cairo/blob/06028d36d098828779f4eda40f5ec7468127c383/crates/cairo-lang-parser/src/parser.rs#L76)):*
> ```XML
>     /// pure-whitespace trivia are cached: they repeat constantly within a file (indentation),
>     /// and a hit skips interning the trivia list into the global green table (hash, shard lock,
>     /// structural equality against a huge map). Sound since green ids are immortal.
> ```

immortality of green ids isn't the interesting part. what makes this sound is that the trivium split is a function of the text - the lexer always splits a whitespace run the same way, and the leading/trailing difference can't yield equal texts with different splits. that's the invariant that silently breaks if the lexer changes, so document that instead.
___
*[`crates/cairo-lang-parser/src/parser.rs` line 3738 at r5](https://reviewable.io/reviews/starkware-libs/cairo/10343#-P-94-dn1wrTytsOT2zq:-P-94-dn1wrTytsOT2zr:b-ck3kt0) ([raw file](https://github.com/starkware-libs/cairo/blob/06028d36d098828779f4eda40f5ec7468127c383/crates/cairo-lang-parser/src/parser.rs#L3738)):*
> ```XML
> 
>         // Pending trivia is source-contiguous and ends exactly at `terminal_start`.
>         let leading_start = terminal_start.sub_width(trivia_total_width(self.db, &pending_trivia));
> ```

`take_raw` just called `LexerTerminal::width`, which sums exactly leading + text + trailing. here we sum leading and text again, and `trivia_green` sums leading and trailing a third time - every `trivium.0.width(db)` is a green interner lookup. that's 2-3x the width traversals per terminal, added by an opt PR. taking the trivia text from the lexer removes all of it.
___
*[`crates/cairo-lang-parser/src/parser.rs` line 3759 at r5](https://reviewable.io/reviews/starkware-libs/cairo/10343#-P-93zsQ4ZBT15yklon9:-P-93zsRBlgQaaDZdM_Z:b-c9prun) ([raw file](https://github.com/starkware-libs/cairo/blob/06028d36d098828779f4eda40f5ec7468127c383/crates/cairo-lang-parser/src/parser.rs#L3759)):*
> ```XML
>     /// its text, which keys the cache (for pure-whitespace trivia - see
>     /// [`Parser::trivia_greens`]).
>     fn trivia_green(&mut self, trivia: &[TriviumGreen<'a>], start: TextOffset) -> TriviaGreen<'a> {
> ```

bugbot flagged the `take_doc` offset bug, but the deeper issue is that the cache key is re-derived from an offset threaded through call sites at all - that invariant is exactly what broke, and the failure is silent green-tree corruption. repro:

```
mod a
<40 spaces>{
// c
fn f() {
     let x = 1;
}
}
```

the header-doc trivia is stored under key `"     "`, and `let`'s indentation then comes back from the cache as `// c\n`. also `TextSpan::take` byte-slices the source, so a misaligned offset panics on non-ascii instead of merely corrupting.

rather than patching `take_doc`, don't re-derive the text: the lexer already has the exact span of every trivia run (`previous_position`..`current_position` in `match_trivia`) - hand it out on `LexerTerminal` and drop `terminal_start`, `leading_start`/`trailing_start` and the `Parser::text` field entirely. and pending trivia always holds a `TokenSkipped`/skipped node, so it is never pure whitespace - that branch needs no cache at all:

```suggestion
        let leading = if pending_trivia.is_empty() {
            self.trivia_green(&leading_trivia, leading_trivia_text)
        } else {
            pending_trivia.extend(leading_trivia);
            Trivia::new_green(self.db, &pending_trivia)
        };
```
___
*[`crates/cairo-lang-parser/src/parser.rs` line 3762 at r5](https://reviewable.io/reviews/starkware-libs/cairo/10343#-P-941IbAtMN_rUBfmn8:-P-941IbAtMN_rUBfmn9:b-735wdj) ([raw file](https://github.com/starkware-libs/cairo/blob/06028d36d098828779f4eda40f5ec7468127c383/crates/cairo-lang-parser/src/parser.rs#L3762)):*
> ```XML
>         let end = start.add_width(trivia_total_width(self.db, trivia));
>         let text = TextSpan::new(start, end).take(self.text);
>         if text.bytes().all(|b| b.is_ascii_whitespace()) {
> ```

empty trivia - every terminal without trailing trivia, i.e. most of them - is vacuously "all whitespace", so it pays a hash + map lookup on `""`. and any zero-width non-empty trivia list silently aliases that same entry. hoist it:

```suggestion
        if trivia.is_empty() {
            return self.empty_trivia;
        }
        if text.bytes().all(|b| b.is_ascii_whitespace()) {
```

with `empty_trivia: TriviaGreen<'a>` built once in `Parser::new`.
___
*[`crates/cairo-lang-parser/src/parser.rs` line 3763 at r5](https://reviewable.io/reviews/starkware-libs/cairo/10343#-P-940Lh5SuhJ2BliRtW:-P-940Lh5SuhJ2BliRtX:bz56cuc) ([raw file](https://github.com/starkware-libs/cairo/blob/06028d36d098828779f4eda40f5ec7468127c383/crates/cairo-lang-parser/src/parser.rs#L3763)):*
> ```XML
>         let text = TextSpan::new(start, end).take(self.text);
>         if text.bytes().all(|b| b.is_ascii_whitespace()) {
>             *self.trivia_greens.entry(text).or_insert_with(|| Trivia::new_green(self.db, trivia))
> ```

no test for any of this. note `SyntaxNode::get_text` reads the source by span, so a wrong trivia green is invisible to it - only a green-tree print catches it, which is why the whole suite passes today with the tree wrong. add a `partial_parser_tree_with_trivia` case: module body whose `{` is deeply indented, followed by a non-doc header comment, plus a later line indented to the same width.

File: crates/cairo-lang-parser/src/parser.rs [hunk 1]
```diff
@@ -3811,9 +3847,11 @@ impl<'a, 'mt> Parser<'a, 'mt> {
             leading_trivia: header_doc,
             trailing_trivia: vec![],
         };
+        let terminal_start = self.offset;
         self.offset = self.offset.add_width(empty_lexer_terminal.width(self.db));
 
-        let empty_terminal = self.add_trivia_to_terminal::<TerminalEmpty<'_>>(empty_lexer_terminal);
+        let empty_terminal =
+            self.add_trivia_to_terminal::<TerminalEmpty<'_>>(empty_lexer_terminal, terminal_start);
```
[the remark this answers is not in the document]
Comment (orizi) [reply crates/cairo-lang-parser/src/parser.rs:3861]: Confirmed and fixed - `terminal_start` is now `self.offset.add_width(self.current_width)` (the start of the next terminal). Added a golden test (`partial_trees_with_trivia/header_doc`) reproducing the exact cache-pollution scenario; it fails without the fix.
_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by orizi** [comment]
Responses to the review comments (Reviewable's agent API has been erroring for hours, so posting here; drafts are queued there as well):

- **Width recomputation** (`add_trivia_to_terminal`): reworked — each trivia list's width is summed once and complete spans are passed to `trivia_green`, removing the traversals this PR added. `take_raw`'s `terminal.width()` is a pre-existing cost on main; moving widths into `LexerTerminal` is a worthwhile follow-up but churns the lexer API and its test constructors.
- **Missing test**: added `partial_trees_with_trivia/header_doc` per the suggested recipe, fed via `>>> file:` since the framework auto-formats inline `cairo_code` (and formatted code cannot reach the broken offsets — which is also why the suite passed with the tree wrong). It reproduces the aliasing exactly and fails without the `take_doc` fix.
- **Empty trivia**: done — `empty_trivia` is built once in `Parser::new` and returned early for empty lists.
- **Cache-soundness doc**: rewritten around the invariant — the lexer's split of a whitespace run into trivia is a pure function of its text, so equal texts always have equal trivia lists.

Also fixed the cursor-reported `take_doc` offset bug (`terminal_start` now `self.offset.add_width(self.current_width)`), which the new golden pins.
_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Review by eytan-starkware — COMMENTED** [review]
> @eytan-starkware+AGNT resolved 5 discussions.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/10343)* status: 0 of 9 files reviewed, all discussions resolved (waiting on TomerStarkware).

**Review by eytan-starkware — CHANGES_REQUESTED** [review]
> @eytan-starkware reviewed 9 files and all commit messages, and made 4 comments.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/10343)* status: all files reviewed, 4 unresolved discussions (waiting on orizi and TomerStarkware).

___
*[`crates/cairo-lang-parser/src/parser.rs` line 76 at r6](https://reviewable.io/reviews/starkware-libs/cairo/10343#-P-O74yk-Y64XDjlMkln:-P-O74yk-Y64XDjlMklo:bogqgq5) ([raw file](https://github.com/starkware-libs/cairo/blob/e4fb9aece6ccfa16698d82690c9c4225372aace9/crates/cairo-lang-parser/src/parser.rs#L76)):*
> ```XML
>     /// pure-whitespace trivia are cached: they repeat constantly within a file (indentation),
>     /// and a hit skips interning the trivia list into the global green table (hash, shard lock,
>     /// structural equality against a huge map). Sound as the lexer's split of a whitespace run
> ```

Too long a comment
___
*[`crates/cairo-lang-parser/src/parser.rs` line 194 at r6](https://reviewable.io/reviews/starkware-libs/cairo/10343#-P-O7D6c0D0-3v_qIiQZ:-P-O7D6c0D0-3v_qIiQ_:b6tafch) ([raw file](https://github.com/starkware-libs/cairo/blob/e4fb9aece6ccfa16698d82690c9c4225372aace9/crates/cairo-lang-parser/src/parser.rs#L194)):*
> ```XML
>             macro_parsing_context: MacroParsingContext::None,
>             trivia_greens: Default::default(),
>             empty_trivia: Trivia::new_green(db, &[]),
> ```

Is empty trivia needed or could it just be the empty list in the trivia_greens?
___
*[`crates/cairo-lang-parser/src/parser.rs` line 3768 at r6](https://reviewable.io/reviews/starkware-libs/cairo/10343#-P-O9nUc46mPolZTVC3r:-P-O9nUc46mPolZTVC3s:bz6ks5) ([raw file](https://github.com/starkware-libs/cairo/blob/e4fb9aece6ccfa16698d82690c9c4225372aace9/crates/cairo-lang-parser/src/parser.rs#L3768)):*
> ```XML
>         }
>         let text = span.take(self.text);
>         if text.bytes().all(|b| b.is_ascii_whitespace()) {
> ```

Please add a test of skipped with differnet recovery paths to show why we need the whitespace limitation or doc.
Also: **One nit worth flagging on the PR:** the gate uses `u8::is_ascii_whitespace`, which accepts form feed (`\x0c`) — but the lexer does _not_ treat form feed as whitespace; it becomes a bad-character terminal that enters trivia via the skipped-token machinery. A span like `" \x0c "` would pass the gate while its trivia list contains a `TokenSkipped`, i.e. exactly the category the gate exists to exclude. In practice it looks deterministic today (bad characters seem to always take the flat `TokenSkipped` route, never a `SkippedNode`), so it's likely benign, but matching the lexer's actual whitespace set — `matches!(b, b' ' | b'\t' | b'\r' | b'\n')` — would make the gate airtight and self-documenting.
___
*[`crates/cairo-lang-parser/src/parser_test_data/partial_trees_with_trivia/header_doc` line 20 at r6](https://reviewable.io/reviews/starkware-libs/cairo/10343#-P-OAIDGBWSLiYC--qai:-P-OAIDGBWSLiYC--qaj:b254nrl) ([raw file](https://github.com/starkware-libs/cairo/blob/e4fb9aece6ccfa16698d82690c9c4225372aace9/crates/cairo-lang-parser/src/parser_test_data/partial_trees_with_trivia/header_doc#L20)):*
> ```
> //! > expected_tree
> └── root (kind: SyntaxFile)
>     ├── items (kind: ModuleItemList)
> ```

Is this test useful? Isnt one white space longer then the other?

**Review by orizi — COMMENTED** [review]
> @orizi made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/10343)* status: all files reviewed, 5 unresolved discussions (waiting on TomerStarkware).

___
*[`crates/cairo-lang-parser/src/parser_test_data/partial_trees_with_trivia/header_doc` line 7 at r6](https://reviewable.io/reviews/starkware-libs/cairo/10343#-P-OVGFfCK-A554ulYax:-P-OVGFfCK-A554ulYay:bx1y3l) ([raw file](https://github.com/starkware-libs/cairo/blob/e4fb9aece6ccfa16698d82690c9c4225372aace9/crates/cairo-lang-parser/src/parser_test_data/partial_trees_with_trivia/header_doc#L7)):*
> ```
> 
> //! > cairo_code
> >>> file: src/parser_test_data/cairo_test_files/header_doc_indent.cairo
> ```

any reason to not keep the test inline?

if for formatting issues - i assume skip fmt attributes should be enough.

**Review by orizi — COMMENTED** [review]
> @orizi+AGNT made 4 comments.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/10343)* status: 7 of 10 files reviewed, 5 unresolved discussions (waiting on eytan-starkware and TomerStarkware).

___
*[`crates/cairo-lang-parser/src/parser.rs` line 76 at r6](https://reviewable.io/reviews/starkware-libs/cairo/10343#-P-O74yk-Y64XDjlMkln:-P-OVygk6WKbYUnNSxo1:b-af05cc) ([raw file](https://github.com/starkware-libs/cairo/blob/e4fb9aece6ccfa16698d82690c9c4225372aace9/crates/cairo-lang-parser/src/parser.rs#L76)):*
<details><summary><i>Previously, eytan-starkware wrote…</i></summary><blockquote>

Too long a comment
</blockquote></details>

Done - trimmed to three lines.
___
*[`crates/cairo-lang-parser/src/parser.rs` line 194 at r6](https://reviewable.io/reviews/starkware-libs/cairo/10343#-P-O7D6c0D0-3v_qIiQZ:-P-OVzCi0kvuHzcSUhJJ:b-1bzs94) ([raw file](https://github.com/starkware-libs/cairo/blob/e4fb9aece6ccfa16698d82690c9c4225372aace9/crates/cairo-lang-parser/src/parser.rs#L194)):*
<details><summary><i>Previously, eytan-starkware wrote…</i></summary><blockquote>

Is empty trivia needed or could it just be the empty list in the trivia_greens?
</blockquote></details>

It could, but the explicit branch keeps two things: most terminals have an empty trivia side, so this skips the hash-and-lookup on `""` for the most common case; and a hypothetical zero-width non-empty trivia list (whose span text is also `""`) can never alias the empty list. Happy to fold it into the map if you prefer the smaller surface.
___
*[`crates/cairo-lang-parser/src/parser.rs` line 3768 at r6](https://reviewable.io/reviews/starkware-libs/cairo/10343#-P-O9nUc46mPolZTVC3r:-P-OVzlG1fKFr4F8jRzc:b-14girl) ([raw file](https://github.com/starkware-libs/cairo/blob/e4fb9aece6ccfa16698d82690c9c4225372aace9/crates/cairo-lang-parser/src/parser.rs#L3768)):*
<details><summary><i>Previously, eytan-starkware wrote…</i></summary><blockquote>

Please add a test of skipped with differnet recovery paths to show why we need the whitespace limitation or doc.
Also: **One nit worth flagging on the PR:** the gate uses `u8::is_ascii_whitespace`, which accepts form feed (`\x0c`) — but the lexer does _not_ treat form feed as whitespace; it becomes a bad-character terminal that enters trivia via the skipped-token machinery. A span like `" \x0c "` would pass the gate while its trivia list contains a `TokenSkipped`, i.e. exactly the category the gate exists to exclude. In practice it looks deterministic today (bad characters seem to always take the flat `TokenSkipped` route, never a `SkippedNode`), so it's likely benign, but matching the lexer's actual whitespace set — `matches!(b, b' ' | b'\t' | b'\r' | b'\n')` — would make the gate airtight and self-documenting.
</blockquote></details>

Done on both: added `partial_trees_with_trivia/skipped`, where the same `#[;]` text is skipped token-by-token in one place (generic params) and as a `SkippedNode` in another (module level) - verified the golden goes red if the whitespace limitation is lifted (the second occurrence gets the first's cached list). And tightened the gate to the lexer's exact whitespace set (space/tab/CR/newline) with a comment on the form-feed pitfall.
___
*[`crates/cairo-lang-parser/src/parser_test_data/partial_trees_with_trivia/header_doc` line 20 at r6](https://reviewable.io/reviews/starkware-libs/cairo/10343#-P-OAIDGBWSLiYC--qai:-P-OW-FcEK0covyGswwz:bre50oq) ([raw file](https://github.com/starkware-libs/cairo/blob/e4fb9aece6ccfa16698d82690c9c4225372aace9/crates/cairo-lang-parser/src/parser_test_data/partial_trees_with_trivia/header_doc#L20)):*
<details><summary><i>Previously, eytan-starkware wrote…</i></summary><blockquote>

Is this test useful? Isnt one white space longer then the other?
</blockquote></details>

It is deliberate: the buggy `terminal_start` points at the start of the `{` terminal's 8-space indentation, and the header trivia (`// abc` + newline) is 7 wide - so the buggy cache key is the first 7 chars of that indentation, i.e. 7 spaces, exactly matching `fn bar`'s 7-space indent later. The one-space difference is what makes the wrongly-read span land inside the indentation. Verified the test is red without the `take_doc` offset fix.

**Review by orizi — COMMENTED** [review]
> @orizi+AGNT made 1 comment and resolved 1 discussion.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/10343)* status: 4 of 10 files reviewed, 4 unresolved discussions (waiting on eytan-starkware and TomerStarkware).

___
*[`crates/cairo-lang-parser/src/parser_test_data/partial_trees_with_trivia/header_doc` line 7 at r6](https://reviewable.io/reviews/starkware-libs/cairo/10343#-P-OVGFfCK-A554ulYax:-P-OY5oTEyICJnPB2F96:b8dk19o) ([raw file](https://github.com/starkware-libs/cairo/blob/e4fb9aece6ccfa16698d82690c9c4225372aace9/crates/cairo-lang-parser/src/parser_test_data/partial_trees_with_trivia/header_doc#L7)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

any reason to not keep the test inline?

if for formatting issues - i assume skip fmt attributes should be enough.
</blockquote></details>

Done - inlined with `#[cairofmt::skip]` on the mod (the framework formats inline `cairo_code` on fix, which normalized the deliberate indentation), and removed the external file + its `.cairofmtignore` entry. Re-verified the test still goes red without the `take_doc` offset fix.

**Review by eytan-starkware — APPROVED** [review]
:lgtm:

> @eytan-starkware reviewed 6 files and all commit messages, made 2 comments, and resolved 4 discussions.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/10343#-:-P-SxW8IC10zV4DEalYT:bnfp4nl)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on TomerStarkware).

___
*[`crates/cairo-lang-parser/src/parser.rs` line 194 at r6](https://reviewable.io/reviews/starkware-libs/cairo/10343#-P-O7D6c0D0-3v_qIiQZ:-P-Sw15Q4AtRNbIOZQnT:bz87j7g) ([raw file](https://github.com/starkware-libs/cairo/blob/e4fb9aece6ccfa16698d82690c9c4225372aace9/crates/cairo-lang-parser/src/parser.rs#L194)):*
<details><summary><i>Previously, orizi+AGNT (Agent AGNT for orizi) wrote…</i></summary><blockquote>

It could, but the explicit branch keeps two things: most terminals have an empty trivia side, so this skips the hash-and-lookup on `""` for the most common case; and a hypothetical zero-width non-empty trivia list (whose span text is also `""`) can never alias the empty list. Happy to fold it into the map if you prefer the smaller surface.
</blockquote></details>

Please do

**Review by TomerStarkware — APPROVED** [review]
:lgtm:

> @TomerStarkware reviewed 4 files and all commit messages, and made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/10343#-:-P-UQ2xc4fKQ5bj7kvEW:bnfp4nl)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on orizi).
